### PR TITLE
Add mypy to continuous integration suite

### LINKED ISSUE #268: Use mypy for static type checking in continuous integration
[Mypy](http://mypy.readthedocs.io/en/latest/) uses function annotations to perform static type checking.  Mypy has the potential to help us track down bugs without even running a program.  Lately we've started to add [type hints](https://www.python.org/dev/peps/pep-0484/) to functions and methods in a few of our subpackages (including atomic, for example) where we sometimes make use of the [typing](https://docs.python.org/3/library/typing.html) package as well.  It would be great if we can use mypy in our suite of tests with Travis CI which would involve updating `.travis.yml` (which is also being edited in #264).

A potential problem is that some return annotations are given as units:
```Python
def func() -> u.eV:
```
I am not sure how mypy will handle that situation.  Some function arguments are going to be annotated with the `Particle` class in order for `@particle_input` to know what to do with them.  That may also make mypy grumpy.  Another thing to worry about is that backwards compatibility is not guaranteed for mypy yet, so we might have to make some further adjustments later on.  I personally am inexperienced with mypy, but [projects such as zulip have added mypy to their Travis CI tests](https://github.com/zulip/zulip/issues/635).

Many thanks to whoever ends up working on this!  (And also thanks to @Cadair for letting us know about mypy!)

### PR DESCRIPTION
**Pull request by namurphy**
## Description

This PR adds [`mypy`](https://mypy.readthedocs.io/en/stable/) to our continuous integration suite. 

## Motivation and context

mypy is the *de facto* static type checker in the pythoniverse. It analyzes source code directly and makes sure that type hint annotations are used consistently throughout the package.

Adopting mypy in our CI will help us:

 - Find bugs and common programming errors.
 - Ensure our type hint annotations are accurate and complete. 
 - Better understand what our functions are doing, especially in complicated situations.

Additional considerations or tradeoffs are:

 - Code with type hint annotations is more verbose.  
 - We would need a good way to introduce type hint annotations to new contributors, such as in the contributor guide and in the comment that gets posted to every PR.  We can point to mypy's typing cheatsheet, and discuss some considerations that are specific to PlasmaPy.  
 - I've found error messages from mypy to...take some getting used to.  Additionally, mypy may point to an error on a particular line, when it's a different line that actually needs fixing.
 - mypy is unable to type check objects that are created dynamically rather than being defined statically in the source code (see also https://github.com/astropy/astropy/issues/15808).
 - Our ability to type check things will be limited when upstream packages are not typed, including `astropy.units`.  We can also add type stub files (like in #2442) for important functionality, or `# type: ignore[error-code]` comments for less important situations.  These would need only include the functionality that we actually use.

## Related issues

Closes #268, which was raised almost six years ago!

In #2424, we added a mypy configuration file that ignores existing errors on a per-file and per-error basis. In #2431, we added a tox environment for running mypy.  This PR adds that tox environment to our suite of continuous integration checks.

We also changed `@validate_quantities` to work with `u.Quantity[u.m]` style comments in #2346, following by actual changes to the annotations in #2356 and #2421.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -158,4 +158,4 @@ commands = python -c 'import plasmapy'
 deps =
     mypy >= 1.8.0
```
Comment (namurphy) [inline tox.ini:159]: I'm wondering if there's an alternative to having to update this manually...  dependabot, maybe?

File: tox.ini [hunk 2]
```diff
@@ -158,4 +158,4 @@ commands = python -c 'import plasmapy'
...
 commands =
-    mypy plasmapy --install-types --non-interactive
+    mypy plasmapy --install-types --non-interactive --show-error-context --show-error-code-links --pretty
```
Comment (namurphy) [inline tox.ini:161]: We'll probably want to play with these flags to figure out which error output works best.

**Comment by namurphy** [comment]
It's still really surprising to see mypy passing in our CI suite!  I mean...we've got it set to ignore ∼700 errors, but still!  😺

File: tox.ini (hunk 1 above)
Comment (ejohnson-96) [reply tox.ini:159]: :thinking:

**Review by ejohnson-96 — APPROVED** [review]
:+1: All looks good

**Review by StanczakDominik — APPROVED** [review]
Looks quite sensible, let's try it out :) Please allow me to do something I haven't done in a while...

**Comment by namurphy** [comment]
Thank you for the reviews!  I'm curious about what @rocco8773 thinks about this too.  I'm really happy to have this added to our CI suite since it's been something I've been hoping to do for several years.  

Enabling mypy will level up the quality of our code and documentation, but my main worry is that it'll be an extra barrier to entry for new contributors.  So, we should keep thinking of our mypy configuration as experimental.  I currently have it set so that new modules are expected to follow mypy's strict settings, but we may want to loosen those settings later on for some modules.

@StanczakDominik — It's super awesome to hear from you and I hope you've been doing really well!

**Pull request merged**